be/src/exprs/table_function/java_udtf_function.cpp

```diff
@@ -0,0 +1,179 @@
+};
+
+Status JavaUDTFState::open() {
+    _class_loader = std::make_unique<ClassLoader>(std::move(_libpath));
+    RETURN_IF_ERROR(_class_loader->init());
+    _analyzer = std::make_unique<ClassAnalyzer>();
+
+    _udtf_class = _class_loader->getClass(_symbol);
+    if (_udtf_class.clazz() == nullptr) {
+        return Status::InternalError(fmt::format("Not found symbol:{}", _symbol));
+    }
+
+    RETURN_IF_ERROR(_udtf_class.newInstance(&_udtf_handle));
+
+    auto* analyzer = _analyzer.get();
+    auto add_method = [&](const std::string& name, jclass clazz, std::unique_ptr<JavaMethodDescriptor>* res) {
+        std::string method_name = name;
+        std::string sign;
+        std::vector<MethodTypeDescriptor> mtdesc;
+        RETURN_IF_ERROR(analyzer->get_signature(clazz, method_name, &sign));
+        RETURN_IF_ERROR(analyzer->get_udaf_method_desc(sign, &mtdesc));
+        *res = std::make_unique<JavaMethodDescriptor>();
+        (*res)->name = std::move(method_name);
+        (*res)->sign = std::move(sign);
+        (*res)->method_desc = std::move(mtdesc);
+        return Status::OK();
+    };
+    RETURN_IF_ERROR(add_method("process", _udtf_class.clazz(), &_process));
+
+    return Status::OK();
+}
+
+Status JavaUDTFFunction::init(const TFunction& fn, TableFunctionState** state) const {
+    std::string libpath;
+    RETURN_IF_ERROR(UserFunctionCache::instance()->get_libpath(fn.fid, fn.hdfs_location, fn.checksum, &libpath));
+    // Now we only support one return types
+    *state = new JavaUDTFState(std::move(libpath), fn.table_fn.symbol, fn.table_fn.ret_types[0]);
+    return Status::OK();
+}
+
+Status JavaUDTFFunction::prepare(TableFunctionState* state) const {
+    // Nothing to do
+    return Status::OK();
+}
+
+Status JavaUDTFFunction::open(TableFunctionState* state) const {
+    RETURN_IF_ERROR(down_cast<JavaUDTFState*>(state)->open());
+    return Status::OK();
+}
+
+Status JavaUDTFFunction::close(TableFunctionState* state) const {
+    delete state;
+    return Status::OK();
+}
+
+std::pair<Columns, ColumnPtr> JavaUDTFFunction::process(TableFunctionState* state, bool* eos) const {
+    Columns res;
+    const Columns& cols = state->get_columns();
+    auto* stateUDTF = down_cast<JavaUDTFState*>(state);
+
+    auto& helper = JVMFunctionHelper::getInstance();
+    JNIEnv* env = helper.getEnv();
+
+    jmethodID methodID = env->GetMethodID(stateUDTF->get_udtf_clazz(), stateUDTF->method_process()->name.c_str(),
+                                          stateUDTF->method_process()->sign.c_str());
+
+    std::vector<jvalue> call_stack;
+    std::vector<jobject> rets;
+    size_t num_rows = cols[0]->size();
+    size_t num_cols = cols.size();
+
+    rets.resize(num_rows);
+    for (int i = 0; i < num_rows; ++i) {
+        for (int j = 0; j < num_cols; ++j) {
+            jvalue val = cast_to_jvalue<true>(stateUDTF->method_process()->method_desc[j + 1], cols[j].get(), j);
+            call_stack.push_back(val);
```

Comment:
can callstack::reserve(num_cols) be invoked to make room beforehand?

Reply:
fixed